proxy.h

```diff
@@ -477,10 +478,10 @@ class sbo_ptr {
   sbo_ptr(sbo_ptr&&) noexcept(std::is_nothrow_move_constructible_v<T>)
       = default;
 
-  T& operator*() { return value_; }
+  T& operator*() const { return value_; }
```

Comment:
question: The one potential downside I see to this is here we have 'physical' `const` but not 'logical' `const` as we're returning a non-const reference by calling a `const` function (meaning you can then mutate the internals of the returned object after calling a `const` function). I've seen this referred to in the past as '`const` broken' (see https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/const.md#classes-of-const-in_correctness for more details). Might it be possible to provide overloads for both `const` and non-`const` versions like so...

```
const T& operator*() const { return value_; }
T& operator*() { return value_; }
```

If this is purely an implementation detail it might be fine but I just wanted to mention it incase.

Reply:
@pr0g 

> The one potential downside I see to this is here we have 'physical' `const` but not 'logical' `const` as we're returning a non-const reference by calling a `const` function (meaning you can then mutate the internals of the returned object after calling a `const` function).

In this case, `value_` is considered as an internal storage of the pointee (as `sbo_ptr` implies "small buffer optimization"). There are many similar practices in various implementations of STL. For example, in MSVC implementation of `std::function`, [getting the underlying object](https://github.com/microsoft/STL/blob/67a96a910fb53aa4bc12801fb9683d30b2647c16/stl/inc/functional#L993-L995) is always `const`, regardless of whether it points to an [internal storage](https://github.com/microsoft/STL/blob/67a96a910fb53aa4bc12801fb9683d30b2647c16/stl/inc/functional#L921). The reason MSVC implementation of `std::function` does not use `mutable` is that it stores an addition pointer inside the object, which is not an optimal memory layout and makes it bigger than `proxy` when the buffer size for SBO is the same.

> Might it be possible to provide overloads for both const and non-const versions like so...
>
> ```cpp
> const T& operator*() const { return value_; }
> T& operator*() { return value_; }
>>> ```

No. Doing this will let `proxy` choose the first overload. Users will not be able to mutate the pointee created from `make_proxy` when SBO is enabled.

> If this is purely an implementation detail it might be fine but I just wanted to mention it incase.

Well noted! Thank you for the comments! Although it is an implementation detail, we still want to keep it as neat as we could😀. Please let us know if there is any further consideration!